examples/custom_search_method/core_search_runner/asha.py

```diff
@@ -0,0 +1,483 @@
+            rung.outstanding_trials -= 1
+
+            for rung_idx in range(0, highest_rung_index + 1):
+                rung = self.asha_search_state.rungs[rung_idx]
+                rung.metrics = list(filter(lambda x: x.request_id != request_id, rung.metrics))
+
+            create = Create(
+                request_id=uuid.uuid4(),
+                hparams=self.sample_params(),
+                checkpoint=None,
+            )
+            ops.append(create)
+            ops.append(
+                ValidateAfter(
+                    request_id=create.request_id,
+                    length=self.asha_search_state.rungs[0].units_needed,
+                )
+            )
+
+            self.asha_search_state.trial_rungs[create.request_id] = 0
+            self.asha_search_state.pending_trials += 1
+
+            return ops
+
+        self.asha_search_state.early_exit_trials.add(request_id)
+        self.asha_search_state.closed_trials.add(request_id)
+        return self._promote_async(request_id, sys.float_info.max)
+
+    ############################################################################
+    # Invoked when the master is asking for the progress of your search method.
+    # Progress is a float number in range [0.0, 1.0], where 1.0 means that
+    # your search methods completed all the trials.
+    # Note that sending progress=1.0 only updates progress in the WebUI,
+    # but it does not close the experiment. Ony sending Shutdown operation can
+    # close the experiment.
+    # In this example, progress computation is based on the number of completed trials
+    # in rung 0 to the total number of trials.
+    def progress(self) -> float:
+        if 0 < self.asha_search_state.max_concurrent_trials < self.asha_search_state.pending_trials:
+            raise RuntimeError("Pending trial is greater than max concurrent trials")
+        all_trials = len(self.asha_search_state.rungs[0].metrics)
+
+        progress = all_trials / (1.2 * self.asha_search_state.max_trials)
+        if all_trials == self.asha_search_state.max_trials:
+            num_valid_trials = (
+                    self.asha_search_state.completed_trials - self.asha_search_state.invalid_trials
+            )
+            progress_no_overhead = num_valid_trials / self.asha_search_state.max_trials
+            progress = max(progress_no_overhead, progress)
+
+        return progress
+
+    ############################################################################
+    # User-defined function that generates hyperparameters for each trial.
+    # The hyperparameters are passed to a trial in the Create operation.
+    # In this example, the model (defined in model_def.py) is expecting the following
+    # hyperparameters:
+    #   -> global_batch_size,
+    #   -> n_filters1
+    #   -> n_filters2,
+    #   -> learning_rate,
+    #   -> dropout,
+    #   -> dropout2.
+    def sample_params(self) -> Dict[str, object]:
```

Comment:
what do you think about having asha take a user defined search space as input so that the search method itself will be agnostic to the actual model training code?  the code below can be moved to perhaps an `utils.py` file and imported int `run_experiment.py`